src/analyses/goto_check.cpp

```diff
@@ -1569,46 +1618,8 @@ void goto_checkt::check_rec(const exprt &expr, guardt &guard)
   else if(expr.id()==ID_member &&
           to_member_expr(expr).struct_op().id()==ID_dereference)
   {
-    const member_exprt &member=to_member_expr(expr);
-    const dereference_exprt &deref=
-      to_dereference_expr(member.struct_op());
-
-    check_rec(deref.pointer(), guard);
-
-    // avoid building the following expressions when pointer_validity_check
-    // would return immediately anyway
-    if(!enable_pointer_check)
-      return;
-
-    // we rewrite s->member into *(s+member_offset)
-    // to avoid requiring memory safety of the entire struct
-    auto member_offset_opt = member_offset_expr(member, ns);
-
-    if(member_offset_opt.has_value())
-    {
-      pointer_typet new_pointer_type = to_pointer_type(deref.pointer().type());
-      new_pointer_type.subtype() = expr.type();
-
-      const exprt char_pointer =
-        typecast_exprt::conditional_cast(
-          deref.pointer(), pointer_type(char_type()));
-
-      const exprt new_address = typecast_exprt(
-        plus_exprt(
-          char_pointer,
-          typecast_exprt::conditional_cast(
-            member_offset_opt.value(), pointer_diff_type())),
-        char_pointer.type());
-
-      const exprt new_address_casted =
-        typecast_exprt::conditional_cast(new_address, new_pointer_type);
-
-      dereference_exprt new_deref{new_address_casted};
-      new_deref.add_source_location() = deref.source_location();
-      pointer_validity_check(new_deref, guard);
-
-      return;
-    }
+    check_rec_member(expr, guard);
+    return;
```

Comment:
This `return` doesn't seem correct to me. I think you should remove it, and also remove the final `forall_operands` from `check_rec_member`.

Reply:
The return is there only to prevent the `forall_operands` loop in case the `check_rec_member` was existed via an explicit `return`. But it seems to be only an optimisation so I did what you asked for.

Reply:
I take it back. Avoiding the recursive call on the member expression operands may be just an optimisation but one of our regression tests (`Malloc23`) requires exactly 4 failing assertions. If we change this section so that the recursion is always called, this test fails.

Reply:
Which ... might be a bugfix? Does it generate more assertions or is it missing some?

Reply:
Doesn't seem so:
```
struct A
{
  int x;
  int y;
};

int main()
{
  struct A *p=malloc(sizeof(int));
  p->x=0; // OK
  p->y=0; // out of bounds
..
```
Unless we skip the recursive call, we would generate a pointer check for `p->ok=0` which, on it's own, would fail.

Reply:
But still the behaviour before and after isn't the same: previously, we only returned in _some_ cases, now we always return early.

Reply:
Yes, but since I included the recursive call on sub-expressions in `rec_check_member` the only thing we skip by returning early is the remaining case analyses none of which can succeed (since we know that `expr.id()==ID_member`).

Reply:
Fair enough, I was just a bit worried that one day someone would add a case. But I suppose testing (or rather: failing tests) should catch that. So it seems fine as is.

Reply:
I thought about it a bit and tried having the `check_rec_member` return a Boolean indicating if the subsequent analysis should be attempted or not: https://github.com/diffblue/cbmc/pull/4611/commits/d6f7f05f2f6c0b86a96dd7a36f132dd1e1b70264.